packages/suite/src/actions/suite/metadataActions.ts

```diff
@@ -946,3 +990,311 @@ export const setEditing = (payload: string | undefined): MetadataAction => ({
+    );
+
+    const [renamedOldEncryptionFiles, oldEncryptionFiles] = arrayPartition(restFiles, file =>
+        file.endsWith(`_v${METADATA.ENCRYPTION_VERSION - 1}.mtdt`),
+    );
+
+    return [currentEncryptionFiles, oldEncryptionFiles, renamedOldEncryptionFiles];
+};
+
+const createMigrationPromise =
+    (
+        entity: LabelableEntity,
+        prevEncryptionVersion: MetadataEncryptionVersion,
+        fetchData: boolean,
+    ) =>
+    async (dispatch: Dispatch, getState: GetState) => {
+        const device = selectDevice(getState());
+        if (!device?.state || !device.metadata[METADATA.ENCRYPTION_VERSION]) {
+            return { success: false, error: 'device unexpected state' };
+        }
+        const prevData =
+            fetchData &&
+            (await dispatch(
+                fetchMetadata({
+                    entity,
+                    encryptionVersion: prevEncryptionVersion,
+                    // tudu huh
+                    provider: getState().metadata.providers[0],
+                }),
+            ));
+
+        const nextKeys = entity[METADATA.ENCRYPTION_VERSION];
+
+        if (!nextKeys) {
+            return { success: false, error: 'next keys are missing' };
+        }
+
+        const dummy = { dummy: getWeakRandomId(getRandomNumberInRange(1, 100)) };
+        const prevKeys = entity[prevEncryptionVersion];
+        if (!prevKeys) {
+            return { success: false, error: 'prev keys are missing' };
+        }
+
+        const defaultEntityData =
+            entity.type === 'account'
+                ? cloneObject(METADATA.DEFAULT_ACCOUNT_METADATA)
+                : cloneObject(METADATA.DEFAULT_WALLET_METADATA);
+        const nextData =
+            prevData && 'data' in prevData ? prevData.data : { ...defaultEntityData, ...dummy };
+
+        const providerInstance = dispatch(
+            getProviderInstance({ clientId: getState().metadata.providers[0]!.clientId }),
+        );
+
+        if (!providerInstance) {
+            // provider should always be set here
+            return { success: false, error: 'provider not connected' };
+        }
+
+        dispatch(
+            setMetadata({
+                ...nextKeys,
+                data: nextData,
+                // todo: huh huh
+                provider: getState().metadata.providers[0]!,
+            }),
+        );
+        const saveResult = await dispatch(
+            encryptAndSaveMetadata({
+                ...nextKeys,
+                data: nextData,
+                // todo: huh huh
+                provider: getState().metadata.providers[0]!,
+            }),
+        );
+
+        if (!saveResult.success) {
+            return saveResult;
+        }
+
+        if (fetchData && saveResult) {
+            // rename only if next version was saved correctly
+            await providerInstance.renameFile(
+                prevKeys.fileName,
+                prevKeys.fileName.replace('.mtdt', '_v1.mtdt'),
+            );
```

Comment:
hmm this means that in case saving next version fail, we probably never rename but migration still ends with success. this is probably not what we want? every single problem during migration should stop it I guess

Reply:
i'd say yes beb350291723b4946fe2e5c9deddfdddfb846ca7

Reply:
hmm on the second thought, I am not sure :D renaming file is not critical. It is only additional optimization aiming at reducing "old files" count.